src/samplers/cpu/cpi/mod.rs

```diff
@@ -91,17 +179,86 @@ impl Sampler for Cpi {
+                            );
+                            continue;
+                        }
+                    }
+                    if counts.time_enabled() != counts.time_running() {}
+                    let cycles = counts[&(group.cycles)];
+                    let instructions = counts[&(group.instructions)];
+
+                    total_cycles += cycles;
+                    total_instructions += instructions;
+                    let tsc = counts[&(group.tsc)];
+                    base_frequency = tsc / running_us as u64;
+                    BASEFREQUENCY.set(base_frequency as i64);
+                    cur_group_data.push(counts);
+                }
+                Err(e) => {
+                    error!("error in sampling the perf group: {e}");
+                }
+            };
+        }
+
+        if cur_group_data.len() == self.groups.len() {
+            // update total cycles and instructions
+            self.total_cycles.set(now, elapsed, total_cycles);
+            self.total_instructions
+                .set(now, elapsed, total_instructions);
+            // update the IPC and IPNS
+            if let Some(prev_group_data) = &self.prev_group_data {
+                let nr_cpu = cur_group_data.len() as i64;
+                if cur_group_data.len() == prev_group_data.len() {
+                    let mut ipc: i64 = 0;
+                    let mut avg_frequency = 0;
+                    let mut ipns: i64 = 0;
+                    for cpu in 0..cur_group_data.len() {
+                        let cur_cycles = cur_group_data[cpu][&(self.groups[cpu].cycles)];
+                        let prev_cycles = prev_group_data[cpu][&(self.groups[cpu].cycles)];
+                        let cur_instructions =
+                            cur_group_data[cpu][&(self.groups[cpu].instructions)];
+                        let prev_instructions =
+                            prev_group_data[cpu][&(self.groups[cpu].instructions)];
+                        let cpu_ipc = (cur_instructions - prev_instructions) * 1000 / (cur_cycles - prev_cycles);
+                        ipc += cpu_ipc as i64;
+                        let cur_mperf = cur_group_data[cpu][&(self.groups[cpu].mperf)];
+                        let prev_mperf = prev_group_data[cpu][&(self.groups[cpu].mperf)];
+                        let cur_aperf = cur_group_data[cpu][&(self.groups[cpu].aperf)];
+                        let prev_aperf = prev_group_data[cpu][&(self.groups[cpu].aperf)];
+                        let ratio =
+                            (cur_aperf - prev_aperf) as f64 / (cur_mperf - prev_mperf) as f64;
+                        let running_frequency = (ratio * base_frequency as f64) as i64;
+                        avg_frequency += running_frequency;
+                        ipns += (cpu_ipc as f64 * ratio) as i64;
+                    }
+                    RUNNINGFREQUENCY.set(avg_frequency / nr_cpu);
+                    IPKC.set(ipc / nr_cpu);
```

Comment:
@brayniac Is this the right way to set the gauge value?

Reply:
Yes. That will overwrite the current gauge reading with the provided value.